be/src/storage/tablet_updates.cpp

```diff
@@ -857,8 +870,75 @@ void TabletUpdates::_apply_rowset_commit(const EditVersionInfo& version_info) {
+                    old_column->append_selective(*old_columns[0], idxes.data(), 0, idxes.size());
+
+                    std::map<uint32_t, std::vector<uint32_t>> new_rowids_by_rssid;
+                    std::vector<uint32_t> rowids;
+                    for (int j = 0; j < upserts[i]->size(); ++j) {
+                        rowids.push_back(j);
+                    }
+                    new_rowids_by_rssid[rowset_id + i] = rowids;
+                    std::vector<std::unique_ptr<vectorized::Column>> new_columns(1);
+                    auto new_column = ChunkHelper::column_from_field_type(tablet_column.type(), tablet_column.is_nullable());
+                    new_columns[0] = new_column->clone_empty();
+                    get_column_values(read_column_ids, false, new_rowids_by_rssid, &new_columns);
+
+                    int idx_begin = 0;
+                    int upsert_idx_step = 0;
+                    for (int  j = 0;  j < old_column->size(); ++ j) {
+                        if (num_default > 0 && idxes[j] == 0) {
+                            // plan_read_by_rssid will return idx with 0 if we have default value
+                            upsert_idx_step++;
+                        } else {
+                            int r = old_column->compare_at(j, j, *new_columns[0].get(), -1);
+                            if (r > 0) {
+                                index.upsert(rowset_id + i, idx_begin, *upserts[i],
+                                             idx_begin, idx_begin + upsert_idx_step, &new_deletes);
+                                manager->index_cache().update_object_size(index_entry, index.memory_usage());
+
+                                idx_begin = j + 1;
+                                upsert_idx_step = 0;
+
+                                // Update delete vector of current segment which is being applied
+                                new_deletes[rowset_id + i].push_back(j);
+                            } else {
+                                upsert_idx_step++;
+                            }
+                        }
+                    }
+
+                    if (idx_begin < old_column->size()) {
+                        index.upsert(rowset_id + i, idx_begin, *upserts[i],
+                                     idx_begin, idx_begin + upsert_idx_step, &new_deletes);
+                        manager->index_cache().update_object_size(index_entry, index.memory_usage());
+                    }
+                } else {
+                    index.upsert(rowset_id + i, 0, *upserts[i], &new_deletes);
+                    manager->index_cache().update_object_size(index_entry, index.memory_usage());
+                }
+            } else {
+                index.upsert(rowset_id + i, 0, *upserts[i], &new_deletes);
+                manager->index_cache().update_object_size(index_entry, index.memory_usage());
+            }
```

Comment:
these 3 conditions all upsert index at end, so it's better to put `prepare upsert value` logic in other place or another method, and call it if `conditional_column>=0`, and leave this part of the code unchanged